### PR TITLE
Refactoring to meet codecov-bash

### PR DESCRIPTION
**Pull request by stevepeak**
Enhance the Python uploader as a global uploader

- [x] add more known coverage file
- [x] Ignores known bad paths
- [x] respects `.bowerrc`
- [x] added arguments
  - [x] `--file` to specify specific files to upload
  - [x] `--dump` will write to stdout collected data and not upload
- [x] enhance logging output for debugging
- [x] fixed various CI environment variables

### CODE REVIEW

**Comment by stevepeak** [comment]
@treyhunner can you review this when you get a change. Thank you :smile:

File: codecov/__init__.py [hunk 1]
```diff
@@ -3,9 +3,9 @@
 import os
 import re
 import sys
-import json
 import requests
 import argparse
+from json import loads
 from json import dumps
```
Comment (treyhunner) [inline codecov/__init__.py]: I would usually combine these two lines like: `from json import dumps, loads`

**Comment by treyhunner** [comment]
@stevepeak there's a lot going on here and I don't understand most of it.  I do see a lot of code that looks like it could be refactored to reduce duplication, but refactoring can always wait.

I did notice you check for `bower_components` and exclude that.  Why?  I didn't see a `node_modules` check.  Would that be relevant too?

I'd be willing to take a stab at some refactoring after merge.  What's the test coverage on this? :smile:

**Comment by stevepeak** [comment]
Thanks @treyhunner for checking it out. I did actually refactor the project just now. The goal is straightforward: `Detect, List, Gather, Upload`.

1. **Detect** CI Provider
2. **List** file structure
  - Most reports present files with full paths ex. `/home/ubuntu/.../.../.../src/etc.` which clearly are not accurate to the Git project. So providing the projects network allows Codecov backend to fix the paths provided in the coverage reports.
3. **Gather** all known coverage reports.
  - There easily over a dozen different coverage reports: types and file names. Codecov accepts them all and in the same upload.
4. **Upload** to Codecov.

**Pull request merged**